### PR TITLE
Prevent NREX when instantiating Java.Lang.StackTraceElement

### PR DESCRIPTION
**Pull request by grendello**
Context: https://github.com/xamarin/xamarin-android/issues/8788#issuecomment-1980765624
Context: https://developer.android.com/reference/java/lang/StackTraceElement?hl=en#StackTraceElement(java.lang.String,%20java.lang.String,%20java.lang.String,%20int)

`Java.Lang.StackTraceElement` constructor requires the `declaringClass`
and `methodName` parameters to never be `null`.  Pass the string
`Unknown` whenever any of this information is missing from the managed
stack frame.

Additionally, the `lineNumber` parameter is now set to `-2` if we think
a stack frame points to native code.

### CODE REVIEW

**Comment by grendello** [comment]
@jonpryor our `Java.Lang.StackTraceElement` MCW generates the following signature which is, IMO, incorrect according to the [Java API docs](https://developer.android.com/reference/java/lang/StackTraceElement?hl=en#StackTraceElement(java.lang.String,%20java.lang.String,%20java.lang.String,%20int)):

```csharp
// Metadata.xml XPath constructor reference: path="/api/package[@name='java.lang']/class[@name='StackTraceElement']/constructor[@name='StackTraceElement' and count(parameter)=4 and parameter[1][@type='java.lang.String'] and parameter[2][@type='java.lang.String'] and parameter[3][@type='java.lang.String'] and parameter[4][@type='int']]"
[Register (".ctor", "(Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;I)V", "")]
public unsafe StackTraceElement (string? declaringClass, string? methodName, string? fileName, int lineNumber) : base (IntPtr.Zero, JniHandleOwnership.DoNotTransfer)
```

Shouldn't `declaringClass` and `methodName` be declared as non-nullable `string`?

**Comment by jpobst** [comment]
> Shouldn't `declaringClass` and `methodName` be declared as non-nullable `string`?

We do not manually assign nullability, we simply pass along what Google has annotated, and it appears this method has no `@NonNull` annotations declared in `android.jar`:

```java
public final class StackTraceElement implements Serializable {
  public StackTraceElement(String declaringClass, String methodName, String fileName, int lineNumber) {
    throw new RuntimeException("Stub!");
  }
  ...
}
```

**Comment by jonpryor** [comment]
@grendello [asked](https://github.com/xamarin/xamarin-android/pull/8795#issuecomment-1983224026):

> Shouldn't `declaringClass` and `methodName` be declared as non-nullable string?

We only flag a parameter or return type as non-null if it contains one of these annotations: https://github.com/xamarin/java.interop/blob/e557044f81917b61031f2ee887700ac43a526475/src/Xamarin.Android.Tools.Bytecode/XmlClassDeclarationBuilder.cs#L503-L512

By default, then, *everything* is allowed to be `null`, which is the behavior you're seeing.

We can use Metadata to update these parameters so that `null` is not allowed.

@jpobst would such an update be a good idea?

```xml
  <attr
      path="/api/package[@name='java.lang']/class[@name='StackTraceElement']/constructor/parameter[@type='java.lang.String']"
      name="not-null">true</attr>
```

**Comment by jpobst** [comment]
> @jpobst would such an update be a good idea?

There are likely thousands of Android APIs that are missing nullability information.  Thus our stated policy is that we will not manually fix them: https://learn.microsoft.com/en-us/xamarin/android/release-notes/11/11.0#monoandroid-nullable-reference-types-compatibility.

**Comment by grendello** [comment]
> > @jpobst would such an update be a good idea?
> 
> There are likely thousands of Android APIs that are missing nullability information. Thus our stated policy is that we will not manually fix them: https://learn.microsoft.com/en-us/xamarin/android/release-notes/11/11.0#monoandroid-nullable-reference-types-compatibility.

While it was fine before nullable support was added to .NET, I think now it would be good to change that policy... I realize it's a lot of work, but this is where the compiler (or the IDE) can help avoid mistakes like this one.

File: src/Mono.Android/Android.Runtime/JavaProxyThrowable.cs [hunk 1]
```diff
@@ -66,15 +66,28 @@ void TranslateStackTrace ()
 			int nElements = frames.Length + (javaTrace?.Length ?? 0);
 			StackTraceElement[] elements = new StackTraceElement[nElements];
 
+			const string Unknown = "Unknown";
 			for (int i = 0; i < frames.Length; i++) {
 				StackFrame managedFrame = frames[i];
 				MethodBase? managedMethod = StackFrameGetMethod (managedFrame);
 
+				// https://developer.android.com/reference/java/lang/StackTraceElement?hl=en#StackTraceElement(java.lang.String,%20java.lang.String,%20java.lang.String,%20int)
+				int lineNumber;
```
Comment (jonpryor) [inline src/Mono.Android/Android.Runtime/JavaProxyThrowable.cs:75]: I'm reminded of: https://github.com/xamarin/xamarin-android/pull/8758/files#r1504920023

wherein `StackFrame.GetMethod()` may return `null`, while `StackFrame.ToString()` contains useful information.

I suspect that we should instead "lean in" to using `StackFrame.ToString()`, as we *know* `managedFrame` is not null (if it were, we'd get a `NullReferenceException` from here!).

Comment (grendello) [reply src/Mono.Android/Android.Runtime/JavaProxyThrowable.cs:75]: I don't want to use `ToString()` output, as it's unwieldy to handle in a presentable manner.  However, the information mentioned above can be obtained using the StackFrame [extension methods](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.stackframeextensions?view=net-8.0) (which this PR already uses in a small capacity)

Comment (jonpryor) [reply src/Mono.Android/Android.Runtime/JavaProxyThrowable.cs:75]: @grendello: parsing `MainActivity.OnCreate()` is more useful than providing `Unknown.Unknown:-1` as the stack frame information to Java.  Is it perfect?  No.  But it's better than the alternative.

**Pull request merged**